dlt/destinations/sql_jobs.py

```diff
@@ -12,36 +12,71 @@
+    """Sql base job for jobs that rely on the whole tablechain"""
+    failed_text: str = ""
 
     @classmethod
     def from_table_chain(cls, table_chain: Sequence[TTableSchema], sql_client: SqlClientBase[Any]) -> NewLoadJobImpl:
-        """Generates a list of sql statements that merge the data in staging dataset with the data in destination dataset.
+        """Generates a list of sql statements, that will be executed by the sql client when the job is executed in the loader.
 
         The `table_chain` contains a list schemas of a tables with parent-child relationship, ordered by the ancestry (the root of the tree is first on the list).
-        The root table is merged using primary_key and merge_key hints which can be compound and be both specified. In that case the OR clause is generated.
-        The child tables are merged based on propagated `root_key` which is a type of foreign key but always leading to a root table.
-
-        First we store the root_keys of root table elements to be deleted in the temp table. Then we use the temp table to delete records from root and all child tables in the destination dataset.
-        At the end we copy the data from the staging dataset into destination dataset.
         """
+
         top_table = table_chain[0]
         file_info = ParsedLoadJobFileName(top_table["name"], uniq_id()[:10], 0, "sql")
         try:
             # Remove line breaks from multiline statements and write one SQL statement per line in output file
             # to support clients that need to execute one statement at a time (i.e. snowflake)
-            sql = [' '.join(stmt.splitlines()) for stmt in cls.gen_merge_sql(table_chain, sql_client)]
+            sql = [' '.join(stmt.splitlines()) for stmt in cls.generate_sql(table_chain, sql_client)]
             job = cls(file_info.job_id(), "running")
             job._save_text_file("\n".join(sql))
         except Exception:
             # return failed job
-            failed_text = "Tried to generate a merge sql job for the following tables:"
             tables_str = yaml.dump(table_chain, allow_unicode=True, default_flow_style=False, sort_keys=False)
             job = cls(file_info.job_id(), "failed", pretty_format_exception())
-            job._save_text_file("\n".join([failed_text, tables_str]))
+            job._save_text_file("\n".join([cls.failed_text, tables_str]))
         return job
 
+    @classmethod
+    def generate_sql(cls, table_chain: Sequence[TTableSchema], sql_client: SqlClientBase[Any]) -> List[str]:
+        pass
+
+
+class SqlStagingCopyJob(SqlBaseJob):
+    """Generates a list of sql statements that copy the data from staging dataset into destination dataset."""
+    failed_text: str = "Tried to generate a staging copy sql job for the following tables:"
+
+    @classmethod
+    def generate_sql(cls, table_chain: Sequence[TTableSchema], sql_client: SqlClientBase[Any]) -> List[str]:
+        sql: List[str] = []
+        for table in table_chain:
```

Comment:
could we do this:
1. redshift, postgres and duckdb have transactional DDL so we could do the following
- DROP destination table
- ALTER TABLE SCHEMA staging -> destination
- CREATE TABLE staging LIKE destination (we must recreate the staging table after move)
2. bigquery
this is really really cool
https://cloud.google.com/bigquery/docs/table-clones-intro

basically
1. DROP destination TABLE if not exist
2. CLONE staging -> destination

bigquery does not have ddl transactions BUT the operation above is idempotent so `dlt` will retry it until it succeeds. IMO good enough

for snowflake we cannot do much -> no DDL transactions. we could rename or swap tables but if we retry the operation will fail. so we should stay with copy

we also need to upgrade `SqlLoadJob` so it detects if sql script contains DDL statements and does not begin transaction if DDL statements are not supported (ie. BigQuery and Motherduck case)

this is some work but IMO it is worthy esp. in case of Redshift and BigQuery we get a huge improvement in cost

Reply:
And the alter table append command in redshift can not run in a tx and can also not be run in the same string of statements as other append commands, so I don't think it makes sense to use it really.